### DISCUSSION TITLE
Supabase JWT for .NET API Auth

### DISCUSSION THREAD

**Discussion by jkergosi** [supabase/supabase#14139] [Questions] [answered]
I have Supabase successfully integrated into my React App, but am trying to use Supabase's JWT for .NET API authentication / authorization.

The JWT is valid (checked in jwt.io), but when I try to use against the API, it consistently gives this error: "Bearer error - invalid_token - The signature key was not found"

My code: 

API call, in Header:	Authorization, Bearer [MY_SUPABASE_JWT]

Startup.cs

services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)

	.AddJwtBearer(options => new JwtBearerOptions
    {
		TokenValidationParameters = new TokenValidationParameters
		{
			ValidateIssuer = true,
            ValidateAudience = true,
            ValidateLifetime = true,
            ValidateIssuerSigningKey = true,
            ValidIssuer = "https://[MY_SUPABASE_PROJECT_ID].supabase.co", 
            ValidAudience = "authenticated",
            IssuerSigningKey = new SymmetricSecurityKey(Encoding.UTF8.GetBytes("[MY_SUPABASE_JWT_SECRET]")) 
        }
    });

MyController.cs

[Authorize]
[HttpPost("query")]
public async Task<List<JObject>> QueryAsync([FromBody] JObject payload)
{
	...
}

Any ideas?

**Comment by omarelshiltawi** [comment]
let me know if this helps https://chat.elapse.ai/conversation/478fa524-cdcd-4974-b218-3bf88d43da0d

**Reply by jkergosi** [reply]
Thanks, Omar. It was helpful, but I don't know if Supabase has a public key in PEM format. They are JWTs. Any idea how I get it?

**Comment by dtaalbers** [comment]
@jkergosi Did you get it work? I am looking for the exact same answer.

**Reply by jkergosi** [reply]
The short answer is no, but I still need this. If you find out anything, please comment here

**Reply by dtaalbers** [reply, marked as answer]
Note that this isn't fully tested yet. So, I don't know if this is the best solution, but it seems to work. Using the JWT token in the `Authorization` header with `Bearer [token]` as value. 

```
 services.AddAuthentication(o =>
            {
                o.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
                o.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
                o.DefaultScheme = JwtBearerDefaults.AuthenticationScheme;
            })
            .AddJwtBearer(o =>
            {
                o.IncludeErrorDetails = true;
                o.SaveToken = true;
                o.TokenValidationParameters = new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(
                        Encoding.UTF8.GetBytes(["JWT SECRET KEY"])
                    ),
                    ValidateIssuer = false,
                    ValidateAudience = true,
                    ValidAudience = "authenticated",
                };
                o.Events = new JwtBearerEvents()
                {
                   //
                };
            });
```

You can get the secret key by running `show app.settings.jwt_secret;` in the SQL editor.

To get, for example the user Id from the JWT, you can do the following in for example an ActionFilter

```
var userId = context.HttpContext.User.Claims.FirstOrDefault(c => c.Type == ClaimTypes.NameIdentifier)?.Value;
```

Or maybe this:

```
var token = await HttpContext.GetTokenAsync(JwtBearerDefaults.AuthenticationScheme, "access_token")
```

**Reply by jkergosi** [reply]
Thanks! It worked.

**Reply by kbegiedza** [reply]
You can find `jwt_secret` in project's dashboard.
`Project Settings > CONFIGURATION > Data API > JWT Settings > JWT Secret`

`show app.settings.jwt_secret;` is not working anymore.
